### PR TITLE
update galley to self-register and reconcile the validatingwebhookconfiguration

### PR DESCRIPTION
**Pull request by ayj**

### CODE REVIEW

File: galley/cmd/galley/cmd/validator.go [hunk 1]
```diff
@@ -57,47 +60,88 @@ func createMixerValidator() (store.BackendValidator, error) {
...
+	}
+	var webhookConfig v1beta1.ValidatingWebhookConfiguration
+	if err := yaml.Unmarshal(webhookConfigData, &webhookConfig); err != nil {
+		return nil, fmt.Errorf("could not decode validatingwebhookconfiguration from %v: %v", webhookConfigFile, err)
+	}
+	caCertPem, err := ioutil.ReadFile(caCertFile)
+	if err != nil {
+		return nil, err
+	}
+	for i := range webhookConfig.Webhooks {
```
Comment (costinm) [inline galley/cmd/galley/cmd/validator.go]: Some validation maybe  ? len(caCertPem) at least - maybe even a x509 parse ?

File: galley/pkg/crd/validation/monitoring.go [hunk 2]
```diff
@@ -42,6 +42,14 @@ var (
 		Name: "galley_validation_http_error",
 		Help: "Resource validation http serve errors",
 	}, []string{"status"})
+	metricWebhookConfigurationUpdateError = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (costinm) [inline galley/pkg/crd/validation/monitoring.go]: Can you check with Doug that the metrics are pulled by prmetheus and have some graph ?

File: galley/pkg/crd/validation/monitoring.go [hunk 3]
```diff
@@ -42,6 +42,14 @@ var (
...
+		Name: "galley_validation_config_update_error",
+		Help: "k8s webhook configuration update error",
+	}, []string{"error"})
+	metricWebhookConfigurationUpdates = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (costinm) [inline galley/pkg/crd/validation/monitoring.go]: The error I reported few times was 'invalid certs' - is it covered by any of the metrics ?

File: pkg/util/webhookpatch.go [hunk 4]
```diff
@@ -65,45 +69,22 @@ func PatchMutatingWebhookConfig(client admissionregistrationv1beta1client.Mutati
...
+		log.Errorf("Failed to retrieve %q validatingwebhookconfigurations: %v", desired.Name, err)
 		return err
 	}
 
-	if string(patch) != "{}" {
-		_, err = client.Patch(webhookConfigName, types.StrategicMergePatchType, patch)
+	updated := current.DeepCopyObject().(*v1beta1.ValidatingWebhookConfiguration)
+	updated.Webhooks = desired.Webhooks
+	if !reflect.DeepEqual(updated, current) {
+		_, err := client.Update(updated)
```
Comment (costinm) [inline pkg/util/webhookpatch.go]: A counter on updates please, and a log. I'm not 100% sure from reading the code that DeepEqual does what it should do ( and will do in all cases ). The object read from k8s may be slightly modified, with added labels, status, etc (just like we do with pods)

File: galley/pkg/crd/validation/monitoring.go (hunk 3 above)
Comment (ayj) [reply galley/pkg/crd/validation/monitoring.go]: Does golang provide a server-side callback for failed tls handshakes? I couldn't find anything when I searched. If not, this type of error needs to be capture by kube-apiserver.

File: galley/pkg/crd/validation/monitoring.go (hunk 2 above)
Comment (ayj) [reply galley/pkg/crd/validation/monitoring.go]: These metrics show up in Prometheus, but they aren't in any of our dashboards. 

@douglas-reid, do we provide istio-specific metrics in our dashboards?

File: galley/cmd/galley/cmd/validator.go [hunk 5]
```diff
@@ -57,47 +60,88 @@ func createMixerValidator() (store.BackendValidator, error) {
...
+	}
+	if err := util.PatchValidatingWebhookConfig(validateClient, desiredValidatingWebhookConfig); err != nil {
+		validation.ReportValidationConfigUpdateError(err)
 		return err
 	}
 
-	watchDir, _ := filepath.Split(caCertFile)
-	if err = watcher.Watch(watchDir); err != nil {
-		return fmt.Errorf("could not watch %v: %v", caCertFile, err)
+	reconcile := func() {
```
[the remark this answers is not in the document]
Comment (ayj) [reply galley/cmd/galley/cmd/validator.go]: Line 82 is dead code. I'll refactor to clean things up a bit.

File: galley/pkg/crd/validation/monitoring.go (hunk 2 above)
Comment (douglas-reid) [reply galley/pkg/crd/validation/monitoring.go]: we have istio-specific metrics in our dashboards, yes. The dash for Mixer, for instance, shows some of the self-monitoring stats from Mixer. We could add the following to the Pilot dash or create a new one.

**Comment by costinm** [comment]
Tests seem to be failing - would be great to get this in next build, can someone check ?

**Comment by ayj** [comment]
```
Error: unknown command "validatingwebhookconfiguration" for "kubectl"
```

Looks like circle is using an older version of kubectl that doesn't support validatingwebhookconfiguration.

File: galley/pkg/crd/validation/config.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		log.Errorf("%v validatingwebhookconfiguration update failed: %v",
+			wh.webhookConfiguration.Name, err)
+		reportValidationConfigUpdateError(err)
+	} else if updated {
+		log.Infof("%v validatingwebhookconfiguration updated", wh.webhookConfiguration.Name)
+		reportValidationConfigUpdate()
+	}
+}
+
+func reconcileWebhookConfigurationHelper(
```
Comment (ozevren) [inline galley/pkg/crd/validation/config.go]: nit: more descriptive names can help readability here:
upsertWebhookConfig, patchOrCreateWebhookConfig etc.

Also comments that explain the high-level logic would be quite useful.

File: tests/e2e/framework/kubernetes.go [hunk 7]
```diff
@@ -431,6 +431,7 @@ func (k *KubeInfo) Teardown() error {
 	if *skipSetup || *skipCleanup {
 		return nil
 	}
+
```
Comment (ozevren) [inline tests/e2e/framework/kubernetes.go]: Nit: extra newline

**Pull request merged**